walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,160 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.Collections;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.Set;
+import java.lang.*;
+import java.util.Comparator;
 
 public final class FindMeetingQuery {
+  private static final int DAY_IN_MINS = (60 * 24);
+
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    // EDGE CASES: request time over/under minutes in a day number of attendees is 0
+    if (request.getDuration() > DAY_IN_MINS || request.getDuration() < 0) {
+      return Arrays.asList();
+    } else if (request.getAttendees() == null && request.getOptionalAttendees().isEmpty()) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+
+    //returns list of needed attendees event times, put together.
+    ArrayList<TimeRange> allEventTimes = getAllEventTimes(request.getAttendees(), events);
+
+    // no attendees match any events
+    if (allEventTimes.isEmpty() && request.getOptionalAttendees().isEmpty()) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    } else {
+      Collections.sort(allEventTimes, TimeRange.ORDER_BY_END);
+    }
+
+    //gets a list of all the possible times for non-optional attendees returns empty if no eventTimes
+    Collection<TimeRange> meetingOptions = getMeetingRequest(allEventTimes, request.getDuration());
+    Collection<TimeRange> optionalMeetingOptions = new ArrayList<TimeRange>();
+    // get optional attendees open times, if open times overlap with set open times 
+    //  then temp. adjust duration and continue
+    if (!request.getOptionalAttendees().isEmpty()) {
+      ArrayList<TimeRange> optionalAttendeeTimes = getAllEventTimes(request.getOptionalAttendees(), events);
+      Collections.sort(optionalAttendeeTimes, TimeRange.ORDER_BY_END);
+      optionalMeetingOptions = getMeetingRequest(optionalAttendeeTimes, request.getDuration());
+
+      if (optionalMeetingOptions.isEmpty() && request.getAttendees().isEmpty()) {
+        return Arrays.asList();
+      } else if (allEventTimes.isEmpty()) {
+        return optionalMeetingOptions;
+      }
+
+      // combine event time lists and return optional times if they exist
+      for (TimeRange optional : optionalAttendeeTimes) {
+        allEventTimes.add(optional);
+      }
+
+      Collections.sort(allEventTimes, TimeRange.ORDER_BY_END);
+      optionalMeetingOptions = getMeetingRequest(allEventTimes, request.getDuration());
+      
+      if (!optionalMeetingOptions.isEmpty()) {
+        return optionalMeetingOptions;
+      }
+    }
+
+    return meetingOptions;
   }
+
+  private Collection<TimeRange> getMeetingRequest(ArrayList<TimeRange> eventTimes, Long duration) {
+    Collection<TimeRange> collection = new ArrayList<TimeRange>();
```

Comment:
a better name than collection would be meetingTimes as collection isn't very descriptive.